model/adjuster/span_id_deduper.go

```diff
@@ -5,103 +5,24 @@
+// DedupeBySpanID returns an adjuster that removes all but one span with the same SpanID.
+// This is useful for when spans are duplicated in archival storage, as happens with
+// ElasticSearch archival.
+func DedupeBySpanID() Adjuster {
 	return Func(func(trace *model.Trace) (*model.Trace, error) {
 		deduper := &spanIDDeduper{trace: trace}
 		deduper.groupSpansByID()
-		deduper.dedupeSpanIDs()
+		deduper.dedupeSpansByID()
 		return deduper.trace, nil
 	})
 }
 
-const (
-	warningTooManySpans = "cannot assign unique span ID, too many spans in the trace"
-)
-
-var maxSpanID = model.NewSpanID(math.MaxUint64)
-
-type spanIDDeduper struct {
-	trace     *model.Trace
-	spansByID map[model.SpanID][]*model.Span
-	maxUsedID model.SpanID
-}
-
-// groupSpansByID groups spans with the same ID returning a map id -> []Span
-func (d *spanIDDeduper) groupSpansByID() {
-	spansByID := make(map[model.SpanID][]*model.Span)
-	for _, span := range d.trace.Spans {
-		if spans, ok := spansByID[span.SpanID]; ok {
-			// TODO maybe return an error if more than 2 spans found
-			spansByID[span.SpanID] = append(spans, span)
-		} else {
-			spansByID[span.SpanID] = []*model.Span{span}
-		}
-	}
-	d.spansByID = spansByID
-}
-
-func (d *spanIDDeduper) isSharedWithClientSpan(spanID model.SpanID) bool {
-	for _, span := range d.spansByID[spanID] {
-		if span.IsRPCClient() {
-			return true
-		}
-	}
-	return false
-}
-
-func (d *spanIDDeduper) dedupeSpanIDs() {
-	oldToNewSpanIDs := make(map[model.SpanID]model.SpanID)
-	for _, span := range d.trace.Spans {
-		// only replace span IDs for server-side spans that share the ID with something else
-		if span.IsRPCServer() && d.isSharedWithClientSpan(span.SpanID) {
-			newID, err := d.makeUniqueSpanID()
-			if err != nil {
-				span.Warnings = append(span.Warnings, err.Error())
-				continue
-			}
-			oldToNewSpanIDs[span.SpanID] = newID
-			span.ReplaceParentID(span.SpanID) // previously shared ID is the new parent
-			span.SpanID = newID
-		}
-	}
-	d.swapParentIDs(oldToNewSpanIDs)
-}
-
-// swapParentIDs corrects ParentSpanID of all spans that are children of the server
-// spans whose IDs we deduped.
-func (d *spanIDDeduper) swapParentIDs(oldToNewSpanIDs map[model.SpanID]model.SpanID) {
-	for _, span := range d.trace.Spans {
-		if parentID, ok := oldToNewSpanIDs[span.ParentSpanID()]; ok {
-			if span.SpanID != parentID {
-				span.ReplaceParentID(parentID)
-			}
-		}
-	}
-}
-
-// makeUniqueSpanID returns a new ID that is not used in the trace,
-// or an error if such ID cannot be generated, which is unlikely,
-// given that the whole space of span IDs is 2^64.
-func (d *spanIDDeduper) makeUniqueSpanID() (model.SpanID, error) {
-	for id := d.maxUsedID + 1; id < maxSpanID; id++ {
-		if _, ok := d.spansByID[id]; !ok {
-			d.maxUsedID = id
-			return id, nil
-		}
+func (d *spanIDDeduper) dedupeSpansByID() {
+	d.trace.Spans = nil
+	for _, spans := range d.spansByID {
+		d.trace.Spans = append(d.trace.Spans, spans[0])
```

Comment:
Hm, I don't think this is a good behavior. There is no guarantee that the spans in storage are in fact duplicates, it's possible that they either have the same ID by mistake (instrumentation bug) or that they represent different stages of readiness of the same span (e.g. some people want to flush spans before they are finished in case the process croaks and the span never makes it to the backend). We had an earlier attempt to formally support partial (or incomplete) spans in #979.

I am not suggesting that we must implement partial spans support in this PR, but silently dropping spans like this is not good either. At minimum I would attempt to add some basic comparison logic (we do have span hashes implemented somewhere), so that if they are indeed identical then this logic is appropriate. If they are not identical, we have these options:
  1. just return all of them (perhaps add a warning to each span "this looks like a duplicate but is different in X")
  2. return the first with a warning that the others were dropped. I prefer **not to take this approach** because it makes it impossible to actually interrogate the data store in order to understand why the spans are not identical.
  3. perform a merge, i.e. the actual partial spans support for #729. The rules for the merge are not trivial
    * if service name or operation name are different, are these still duplicates or completely different spans?
    * how to merge timestamps and durations
    * merge tags (this should be reasonably straightforward)
    * merge logs (tbd how)

Reply:
Do you have a pointer to the span hashes?

Reply:
Ah, I just found model/hash.go and that Span implements Hashable.

Do you think it would be too big a performance hit to hash all the spans and then dedupe by hash?  It's definitely the most straightforward implementation.

Reply:
At query time it's not that bad, and we already do that at write time anyway.